staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+)
+
+// RetryConditionFn is a convenience type used for wrapping a retry condition for WithRetryAfter filter.
+type RetryConditionFn func() (bool, func(w http.ResponseWriter), string)
+
+// WithRetryAfter rejects any incoming new request(s) with a 429 if one of the provided condition holds
+//
+// It includes new request(s) on a new or an existing TCP connection
+// Any new request(s) arriving after a condition fulfills
+// are replied with a 429 and the following response headers:
+//   - 'Retry-After: N` (so client can retry after N seconds, hopefully on a new apiserver instance), where N is defined as [4, 12)
+//   -  any optional headers set by a condition function
+//
+// However, some requests are considered special and always pass to the next handler in the chain.
+// As of today, those requests are the ones that:
+// - match on the excluded paths
+// - originate from the loop back client
+func WithRetryAfter(handler http.Handler, conditions []RetryConditionFn, excludedPaths []string, authorizerAttributesFunc func(ctx context.Context) (authorizer.Attributes, error)) http.Handler {
```

Comment:
why authorizerAttributesFunc ? Is there variation ?

Reply:
Is there even a use-case for multiple conditions? I don't see any. It would allow to split `RetryConditionFn` into a pure predicate and split appear the mutator into another arg for `WithRetryAfter`.